sips/sip-015/sip-015-network-upgrade.md

```diff
@@ -0,0 +1,1644 @@
+depending on whether or not the third `(string-ascii 40)` argument is given.
+
+This function returns a `Response`. On success, the `ok` value is a `Principal`.
+The `err` value is a value tuple with the form `{ error_code: uint, value: (optional principal) }`.
+
+If the single-byte `version-byte` is in the valid range `0x00` to `0x1f`, but is not an appropriate
+version byte for the current network, then the error will be `u0`, and `value` will contain
+`(some principal)`, where the wrapped value is the principal.  If the `version-byte` is not in this range, 
+however, then the `value` will be `none`.
+
+If the `version-byte` is a `buff` of length 0, if the single-byte `version-byte` is a
+value greater than `0x1f`, or the `hash-bytes` is a `buff` of length not equal to 20, then `error_code`
+will be `u1` and `value` will be `None`.
+
+If a name is given, and the name is either an empty string or contains ASCII characters
+that are not allowed in contract names, then `error_code` will be `u2`.
+
+**Rationale:** This method empowers developers to convert between Bitcoin and
+Stacks addresses, and between Stacks addresses and subnet addresses (with
+applications to enabling users to prove control of assets on different chains
+within a smart contract).  In addition, it provides introspection for smart
+contract addresses.
+
+**Examples:**
+
+```clarity
+(principal-construct 0x1a 0xfa6bf38ed557fe417333710d6033e9419391a320) ;; Returns (ok ST3X6QWWETNBZWGBK6DRGTR1KX50S74D3425Q1TPK)
+(principal-construct 0x1a 0xfa6bf38ed557fe417333710d6033e9419391a320 "foo") ;; Returns (ok ST3X6QWWETNBZWGBK6DRGTR1KX50S74D3425Q1TPK.foo)
+(principal-construct 0x16 0xfa6bf38ed557fe417333710d6033e9419391a320) ;; Returns (err (tuple (error_code u0) (value (some SP3X6QWWETNBZWGBK6DRGTR1KX50S74D3433WDGJY))))
+(principal-construct 0x16 0xfa6bf38ed557fe417333710d6033e9419391a320 "foo") ;; Returns (err (tuple (error_code u0) (value (some SP3X6QWWETNBZWGBK6DRGTR1KX50S74D3433WDGJY.foo))))
+(principal-construct 0x   0xfa6bf38ed557fe417333710d6033e9419391a320) ;; Returns (err (tuple (error_code u1) (value none)))
+(principal-construct 0x16 0xfa6bf38ed557fe417333710d6033e9419391a3)   ;; Returns (err (tuple (error_code u1) (value none)))
+(principal-construct 0x20 0xfa6bf38ed557fe417333710d6033e9419391a320) ;; Returns (err (tuple (error_code u1) (value none)))
+(principal-construct 0x1a 0xfa6bf38ed557fe417333710d6033e9419391a320 "") ;; Returns (err (tuple (error_code u2) (value none)))
+(principal-construct 0x1a 0xfa6bf38ed557fe417333710d6033e9419391a320 "foo[") ;; Returns (err (tuple (error_code u2) (value none)))
+```
+
+### New method `get-burn-block-info?`
+
+* **Input Signature:** `(get-burn-block-info? (prop-name BurnBlockPropertyName) (block-height uint))`
+* **Output Signature:** `(optional buff) | (optional (tuple (addrs (list 2 (tuple (hashbytes (buff 32)) (version (buff 1))))) (payout uint)))`
+
```

Comment:
It is problematic that this function returns conflicting output types depending on the property name. It should preferably only take `block-height` as argument and return both the header-hash and the pox-address as a composite value, such as:

```clarity
(get-burn-block-info? u677050)
=>
(some {header-hash: 0xe67141016c88a7f1203eca0b4312f2ed141531f59303a1c267d7d83ab6b977d8,
       addrs: (list {hashbytes: 0x395f3643cea07ec4eec73b4d9a973dcce56b9bf1, version: 0x00}
                    {hashbytes: 0x7c6775e20e3e938d2d7e9d79ac310108ba501ddb, version: 0x01})
       payout: u123})
```

Reply:
I don't think so.  The property name informs the VM of which type signature to use; the fact that the property name is distinct can really be thought of as just syntactic sugar.

`get-block-info?` already behaves this way.  Having `get-burn-block-info?` follow this precedent was considered by the authors to be the least-surprising behavior.

Reply:
Alternatively, follow the single resposibility principle by having separate functions:

```clarity
(burn-block-header-hash (block-height uint))
=> (optional buff)

(burn-block-pox-addrs (block-height uint))
=> (optional (tuple (addrs (list 2 (tuple (hashbytes (buff 32)) (version (buff 1))))) (payout uint)))
```

These also have the benefit of being pure functions rather than adding a new special function.

Reply:
> These also have the benefit of being pure functions rather than adding a new special function.

Why is that a concern?  Clarity already uses non-pure functions everywhere, and that hasn't been an inhibitor to its comprehensibility and applicability.  Developers don't seem to be upset about this; what _does_ make them upset is how difficult it currently is to read and write to Bitcoin (and the Clarity changes in this SIP are meant to make progress on this front).